proposals/0631-main-return-types.rst

```diff
@@ -0,0 +1,214 @@
+Set program exit code by main return type
+=========================================
+
+.. author:: Shea Levy
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/631>`_.
+.. sectnum::
+.. contents::
+
+To reduce surprise and match common programming language behavior,
+this proposal would allow programs to signal their exit code
+based on the return value from ``main``.
+
+Motivation
+----------
+Many languages, perhaps most notably ``C``, allow specifying a program's exit
+code by returning an appropriate value from ``main``. ``Haskell98`` *does*
+allow a ``main`` which yields values of any type, but surprisingly it does not
+do anything with the result: any successful execution will result in a successful
+program exit, even if ``main`` yields ``1`` or ``ExitFailure 1``.
+
+Before this proposal, the following program would exit with code ``0`` even
+when there are no results:
+
+::
+
+ main :: IO ExitCode
+ main = do
+   results <- doSomeWork
+   case results of
+     [] -> pure (ExitFailure 1)
+     _ -> print results >> pure ExitSuccess
+
+With this proposal in place, the program would exit with code ``1`` in that
+case.
+
+This proposal follows `Rust <https://doc.rust-lang.org/std/process/trait.Termination.html>`_
+in using a typeclass for return types which can signal exit codes.
+
+This is a minor quality of life improvement, but one that I've seen hit
+newcomers now and then and matches reasonable expectations of program
+behavior.
+
+Proposed Change Specification
+-----------------------------
+
+If, for a given program, ``(main >>= exitWith . GHC.IO.Exit.report) :: IO ()`` type-checks,
+then the resulting program will behave as if that had been written for ``main``
+instead.
```

Comment:
What happens if it doesn't typecheck?

Reply:
Is it acceptable to say “otherwise, the program behaves as it would have had this proposal not been written”? I don’t know of a formal specification of `main`’s current behavior, which presumably would have to include what the runtime does when exceptions are thrown etc.?

Reply:
It's acceptable. I'm not sure that it's a good idea. Maybe record the option of failing when `IO A` and `A` doesn't have a `Termination` instance as an alternative.

---

From a pure theoretical property, both alternatives can be implemented in the other by means of an overlappable instance:

Old behaviour when no instance:
```haskell
instance {-# OVERLAPPABLE #-} Termination a where
  report _ = ExitSuccess
```

Error when no instance
```haskell
instance {-# OVERLAPPABLE #-} TypeError <some error message> => Termination a where
  report _ = error "unreachable"
```

So the question of which alternative we want is, beyond implementation considerations, largely a question of what we consider the best UX.

Reply:
Addressed